theano/gof/opt.py

```diff
@@ -2948,6 +2949,52 @@ def copy_stack_trace(from_var, to_var):
         to_var.tag.trace = getattr(to_var.tag, 'trace', []) + tr
 
 
+def with_stack_trace(from_var, to_var):
+    """
+    Copies the stack trace from one or more tensor variables to
+    one or more tensor variables and returns the destination variables.
+
+    Parameters
+    ----------
+    from_var
+        Tensor variable or list of tensor variables to copy stack traces from.
+    to_var
+        Tensor variable or list of tensor variables to copy stack traces to.
+
+    Returns
+    -------
+    tensor variable or list of tensor variables
+        `to_var`, augmented with the stack traces from `from_var`.
+
+    Notes
+    -----
+    The stacktrace is assumed to be of the form of a list of lists
+    of tuples. Each tuple contains the filename, line number, function name
+    and so on. Each list of tuples contains the truples belonging to a
+    particular variable.
+
+    """
+    copy_stack_trace(from_var, to_var)
+    return to_var
+
+
+@contextlib.contextmanager
+def inherit_stack_trace(from_var):
+    """
+    Contextmanager that copies the stack trace from one or more tensor variables to all tensor
+    variables constructed in the body.
+
+    Parameters
+    ----------
+    from_var
+        Tensor variable or list of tensor variables to copy stack traces from.
+
+    """
+    with graph.nodes_constructed() as new_nodes:
+        yield
+    copy_stack_trace(from_var, new_nodes)
```

Comment:
Do you mean that new_nodes is defined after the with statement?

Reply:
Yes, this is what make this code work. Otherwise it would raise an error that new_nodes isn't defined.

Python 3 changed the scope of list comprehensions. I didn't know what the scope was for with statement or if it changed between Python2 and 3. I checked all is fine. It didn't changed and the var defined on the line of the with statement still exist after the with statement.

Reply:
I see. Thanks for the explanations.